opentracing-apache-httpclient/src/main/java/io/opentracing/contrib/apache/TracingClientExec.java

```diff
@@ -0,0 +1,151 @@
+   * Tracing {@link ClientExecChain} is executed after redirect exec, so on redirects it is called
+   * multiple times. This is used as an id for {@link HttpClientContext#setAttribute(String, Object)}
+   * to store number of redirects.
+   */
+  protected static final String REDIRECT_COUNT = TracingHttpClientBuilder.class.getName() + ".redirectCount";
+
+  private final RedirectStrategy redirectStrategy;
+  private final ClientExecChain requestExecutor;
+  private final boolean redirectHandlingDisabled;
+
+  private final Tracer tracer = GlobalTracer.get();
+  private final SpanManager spanManager = DefaultSpanManager.getInstance();
+  private final List<ApacheClientSpanDecorator> spanDecorators;
+
+  public TracingClientExec(ClientExecChain clientExecChain, RedirectStrategy redirectStrategy,
+      boolean redirectHandlingDisabled, List<ApacheClientSpanDecorator> spanDecorators) {
+    this.requestExecutor = clientExecChain;
+    this.redirectStrategy = redirectStrategy;
+    this.redirectHandlingDisabled = redirectHandlingDisabled;
+    this.spanDecorators = new ArrayList<>(spanDecorators);
+  }
+
+  @Override
+  public CloseableHttpResponse execute(HttpRoute route,
+      HttpRequestWrapper request,
+      HttpClientContext clientContext,
+      HttpExecutionAware execAware) throws IOException, HttpException {
+
+    Span localSpan = clientContext.getAttribute(ACTIVE_SPAN, Span.class);
+    CloseableHttpResponse response = null;
+    try {
+      if (localSpan != null) {
+        return (response = handleNetworkProcessing(localSpan, route, request, clientContext, execAware));
+      }
+
+      localSpan = handleLocalSpan(request, clientContext);
+      return (response = handleNetworkProcessing(localSpan, route, request, clientContext, execAware));
+    } catch (Exception e) {
+      localSpan.finish();
+      throw e;
+    } finally {
+      if (response != null) {
+        /**
+         * This exec runs after {@link org.apache.http.impl.execchain.RedirectExec} which loops
+         * until there is no redirect or reaches max redirect count.
+         * {@link RedirectStrategy} is used to decide whether localSpan should be finished or not.
+         * If there is a redirect localSpan is not finished and redirect is logged.
+         */
+        Integer redirectCount = clientContext.getAttribute(REDIRECT_COUNT, Integer.class);
+        if (!redirectHandlingDisabled &&
+            clientContext.getRequestConfig().isRedirectsEnabled() &&
+            redirectStrategy.isRedirected(request, response, clientContext) &&
+            ++redirectCount < clientContext.getRequestConfig().getMaxRedirects()) {
+
+          clientContext.setAttribute(REDIRECT_COUNT, redirectCount);
+        } else {
+          localSpan.finish();
+        }
+      }
+    }
+  }
+
+  protected Span handleLocalSpan(HttpRequest httpRequest, HttpClientContext clientContext) {
+    Tracer.SpanBuilder spanBuilder = tracer.buildSpan(httpRequest.getRequestLine().getMethod())
```

Comment:
You are correct this instrumentation will produce always >= 2 spans. 

Generally speaking, there are more ways how to model this. It also highly depends on the hooks what a given framework provides. What also plays a role is the size of created tracing data. There are pros and cons for both, however, I think the most important is that timing data and references are correct.

I was thinking about one span + additional spans for redirects, but there is a problem how spans are referenced:
 
If there is a redirect the first span should contain tags related to the first request(so it represents first request by data and timing information). Then after response we know that there is a redirect, so we know there will be next span, we also know that we should finish the first span because it's end of that execution chain. Then If we immediately create a new span at the end of chain timing related to the previous span is different (assumption that for the same operation span is created at the same place and wraps execution of the same code). If we wait and create redirect span at the beginning of the chain we can use only `followsFrom` because the previous span has been finished.